tt_eager/tt_dnn/op_library/sdpa/multi_core/sdpa_decode_op_multi_core.cpp

```diff
@@ -0,0 +1,628 @@
+    log_debug("dht_granularity: {}", dht_granularity);
+    log_debug("log2_dht_granularity: {}", log2_dht_granularity);
+
+    // Create circular buffers
+
+    tt::DataFormat q_df = tt_metal::datatype_to_dataformat_converter(input_tensor_q.get_dtype());
+    tt::DataFormat k_df = tt_metal::datatype_to_dataformat_converter(input_tensor_k.get_dtype());
+    tt::DataFormat v_df = tt_metal::datatype_to_dataformat_converter(input_tensor_v.get_dtype());
+    tt::DataFormat mask_df = attn_mask.has_value() ? tt_metal::datatype_to_dataformat_converter(attn_mask.value().get_dtype()) : tt::DataFormat::Float16_b;
+    tt::DataFormat out_df = tt_metal::datatype_to_dataformat_converter(output_tensor.get_dtype());
+    tt::DataFormat scalar_df = tt::DataFormat::Float16_b;
+    tt::DataFormat im_df = fp32_dest_acc_en ? tt::DataFormat::Float32 : tt::DataFormat::Float16_b;
+    // tt::DataFormat im_df = tt::DataFormat::Float16_b;
+    tt::DataFormat stats_df = tt::DataFormat::Float16_b;
+
+    uint32_t q_tile_size = tt_metal::detail::TileSize(q_df);
+    uint32_t k_tile_size = tt_metal::detail::TileSize(k_df);
+    uint32_t v_tile_size = tt_metal::detail::TileSize(v_df);
+    uint32_t mask_tile_size = attn_mask.has_value() ? tt_metal::detail::TileSize(mask_df) : 0;
+    uint32_t out_tile_size = tt_metal::detail::TileSize(out_df);
+    uint32_t scalar_tile_size = tt_metal::detail::TileSize(scalar_df);
+    uint32_t im_tile_size = tt_metal::detail::TileSize(im_df);
+    uint32_t stats_tile_size = tt_metal::detail::TileSize(stats_df);
+    uint32_t intermed_output_tiles = (out0_t + 2*PNHt)*(num_cores_per_batch-1);
+
+    log_debug("q_data_format: {}", q_df);
+    log_debug("k_data_format: {}", k_df);
+    log_debug("v_data_format: {}", v_df);
+    log_debug("mask_data_format: {}", mask_df);
+    log_debug("out_data_format: {}", out_df);
+    log_debug("scalar_data_format: {}", scalar_df);
+    log_debug("intermediate_data_format: {}", im_df);
+    log_debug("statistics_data_format: {}", stats_df);
+
+    // CBs
+    // Q input
+    auto c_in0_config = CircularBufferConfig(q_tiles * q_tile_size, {{CB::c_in0, q_df}}).set_page_size(CB::c_in0, q_tile_size);
+    auto cb_in0_id = CreateCircularBuffer(program, core_grid, c_in0_config);
+
+    // K input
+    auto c_in1_config = CircularBufferConfig(k_tiles * k_tile_size, {{CB::c_in1, k_df}}).set_page_size(CB::c_in1, k_tile_size);
+    auto cb_in1_id = CreateCircularBuffer(program, core_grid, c_in1_config);
+    // V input
+    auto c_in2_config = CircularBufferConfig(v_tiles * v_tile_size, {{CB::c_in2, v_df}}).set_page_size(CB::c_in2, v_tile_size);
+    auto cb_in2_id = CreateCircularBuffer(program, core_grid, c_in2_config);
+
+    // attn_mask input
+    auto c_in3_config = CircularBufferConfig(mask_tiles * mask_tile_size, {{CB::c_in3, mask_df}}).set_page_size(CB::c_in3, mask_tile_size);
+    auto cb_in3_id = CreateCircularBuffer(program, core_grid, c_in3_config);
```

Comment:
You have a lot of CBs here. I would add all of them as compile-time args. It makes code changes later easier.

Reply:
could you please elaborate on that? How do you add them as compile-time args? It would be great if you can point me to an example. Thanks!

Reply:
Yea, we should always do this for generality. There's two useful outcomes:
- sometimes kernels are shared between ops and the ops may setup different CBs for inputs. This allows it to be reused more conveniently
- if you have a lot of CBs, it's a lot easier to add/update CBs. Sometimes you might want to add a new intermediate CB in the middle of the existing ones (because logically it makes more sense) and it would be a pain to change all the hard-coded values.